File: 0496-next-greater-element-i/solution.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> s;
        unordered_map<int,int> m;
        for(auto ele : nums2){
            while(!s.empty() && ele > s.top()){
                m[s.top()] = ele;
                s.pop();
            }
            s.push(ele);
        }
        while(!s.empty()){
            m[s.top()]=-1;
            s.pop();
        }
        vector<int> ans;
        for(auto ele : nums1){
            ans.push_back(m[ele]);
        }
        return ans;
    }
};

File: 0496-next-greater-element-i/solution.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int n = nums2.length;
        ArrayList<Integer> arr = new ArrayList<>(Collections.nCopies(n,-1));
        Stack<Integer> s = new Stack<>();
        for(int i =0;i<n;i++){
            while(!s.isEmpty() && nums2[i] > nums2[s.peek()]){
                int idx = s.pop();
                arr.set(idx,nums2[i]);
            }
            s.push(i);
        }
        int n2 = nums1.length;
        int[] ans = new int[n2];
        for(int i = 0;i<n2;i++){
            for(int j = 0;j<n;j++){
                if(nums1[i] == nums2[j]){
                    ans[i] = arr.get(j);
                }
            }
        }
        return ans;
    }
}
